File: app/src/main/java/com/example/vlad/financemanager/ui/activities/MoneyCalculatorActivity.java
package com.example.vlad.financemanager.ui.activities;

import android.annotation.SuppressLint;
import android.app.DatePickerDialog;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.example.vlad.financemanager.data.database.DatabaseHelper;
import com.example.vlad.financemanager.data.mappers.SpinnerItemMapper;
import com.example.vlad.financemanager.data.models.Account;
import com.example.vlad.financemanager.data.models.Category;
import com.example.vlad.financemanager.ui.IMoneyCalculation;
import com.example.vlad.financemanager.PresenterMoneyCalculator;
import com.example.vlad.financemanager.R;
import com.example.vlad.financemanager.data.models.Operation;
import com.example.vlad.financemanager.data.models.SpinnerItem;
import com.example.vlad.financemanager.ui.adapters.ImageSpinnerAdapter;
import com.example.vlad.financemanager.ui.fragments.DatePickerFragment;

import java.math.BigDecimal;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.OnTextChanged;

public class MoneyCalculatorActivity extends AppCompatActivity implements IMoneyCalculation,
        DatePickerDialog.OnDateSetListener {
    public static final String DATE_KEY = "date_key";
    private static final String DATE_PICKER_TAG = "date picker";

    private final SimpleDateFormat sdf = new SimpleDateFormat("E, dd MMMM");
    private final SimpleDateFormat sdfWithYear = new SimpleDateFormat("E, MMMM dd, yyyy");

    @BindView(R.id.calculationResultTextView) TextView resultText;
    @BindView(R.id.amountMoneyActivityEditText) EditText amountEditText;
    @BindView(R.id.commentMoneyActivityEditText) EditText comment;
    @BindView(R.id.calculatorActivityToolbar) Toolbar toolbar;
    @BindView(R.id.toolbarTitleTextView) TextView toolbarTitle;
    @BindView(R.id.accountSpinner) Spinner accountsSpinner;
    @BindView(R.id.categorySpinner) Spinner categoriesSpinner;
    @BindView(R.id.operationDateButton) Button dateButton;
    @BindView(R.id.saveRecordButton) Button saveButton;
    @BindView(R.id.calculatorBackButton) ImageButton btnBack;

    private PresenterMoneyCalculator presenter;
    private Date operationDate = new Date();
    private DatabaseHelper databaseHelper;

    List<SpinnerItem> categorySpinnerItemList;
    List<SpinnerItem> accountSpinnerItemList;

    private boolean isOperationIncome;
    private int accountId;
    private int categoryId;

    @SuppressLint("ResourceAsColor")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_money_calculator);
        ButterKnife.bind(this);

        databaseHelper = DatabaseHelper.Companion.getInstance(getApplicationContext());
        presenter = new PresenterMoneyCalculator(this);

        btnBack.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                presenter.clearNumber();
                return true;
            }
        });

        initAmountEditText();
        initUiViaExtras(getIntent().getExtras());
    }

    private void initAmountEditText() {
        amountEditText.requestFocus();
        amountEditText.setOnKeyListener(new View.OnKeyListener() {
            @Override
            public boolean onKey(View view, int keyCode, KeyEvent keyEvent) {
                if (!((keyEvent.getAction() == KeyEvent.ACTION_DOWN) && (keyCode == KeyEvent.KEYCODE_ENTER))) {
                    presenter.calculatorReset();
                }

                return false;
            }
        });
    }



    private void initSpinnersItemLists(int userId) {
        List<Account> accountList = databaseHelper.getAllAccounts(userId);
        accountSpinnerItemList = SpinnerItemMapper.mapAccountsToSpinnerItems(accountList);
        List<Category> categoryList = databaseHelper.getAllCategories(userId, isOperationIncome);
        categorySpinnerItemList = SpinnerItemMapper.mapCategoryToSpinnerItems(categoryList);
    }

    private void initToolbar(Boolean isIncome) {
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        toolbar.getNavigationIcon().setColorFilter(Color.WHITE, PorterDuff.Mode.SRC_ATOP);
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });

        String title = isIncome ? getString(R.string.income) : getString(R.string.outcome);
        toolbarTitle.setText(title);
    }

    private void initDateTimePicker(final Date operationDate) {
        dateButton.setText(sdf.format(new Date()));
        dateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DatePickerFragment df = new DatePickerFragment();
                df.setCalendar(operationDate);
                df.show(getSupportFragmentManager(), DATE_PICKER_TAG);
            }
        });
    }

    private String getDateButtonTitleByDate(Long dateInMillis) {
        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(dateInMillis);
        operationDate = calendar.getTime();

        String dateButtonTitle;
        if (calendar.get(Calendar.YEAR) != Calendar.getInstance().get(Calendar.YEAR)) {
            dateButtonTitle = sdfWithYear.format(calendar.getTime());
        } else {
            dateButtonTitle = sdf.format(calendar.getTime());
        }

        return dateButtonTitle;
    }

    @Override
    public int getCategoryId() {
        return categoryId;
    }

    @Override
    public int getAccountId() {
        return accountId;
    }

    @Override
    public Date getOperationDate() {
        return operationDate;
    }

    @Override
    public String getComment() {
        return comment.getText().toString();
    }

    @Override
    public boolean getIsOperationInput() {
        return isOperationIncome;
    }

    @Override
    public String getAmount() {
        return amountEditText.getText().toString();
    }

    @Override
    public void setAmountResultText(String result) {
        amountEditText.setText(result);
        resultText.setText(result);
    }

    @Override
    public void setCalculatorToZero() {
        resultText.setText("0");
    }

    public void calculatorBtnOnClick(View view) {
        Button pressedButton = (Button) view;
        presenter.calculatorBtnOnClick(pressedButton.getId(), pressedButton.getText().toString());
    }

    public void finishActivity() {
        finish();
    }

    public void calculationErrorSignal() {
        Toast.makeText(this, getString(R.string.incorrect_input), Toast.LENGTH_SHORT).show();
    }

    public void calculationErrorSignal(String msg) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void sendNewOperation(Operation operation) {
        Intent intent = new Intent();
        String amount = operation.getAmount().toString();
        Bundle extras = new Bundle();
        extras.putString(MainActivity.AMOUNT_KEY, amount);
        extras.putSerializable(MainActivity.OPERATION_KEY, operation);
        intent.putExtras(extras);

        setResult(0, intent);
    }

    @Override
    public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.YEAR, year);
        calendar.set(Calendar.MONTH, month);
        calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);

        operationDate = calendar.getTime();
        dateButton.setText(getDateButtonTitleByDate(operationDate.getTime()));
    }

    public void initUiViaExtras(Bundle extras) {
        int userId = extras.getInt(MainActivity.USER_ID_KEY);
        isOperationIncome = extras.getBoolean(MainActivity.IS_OPERATION_INCOME);
        boolean isModifyingOperation = extras.getBoolean(MainActivity.IS_MODIFYING_OPERATION);

        initSpinnersItemLists(userId);
        initSpinnersWithItemLists(accountSpinnerItemList, categorySpinnerItemList);
        initToolbar(isOperationIncome);

        if (isModifyingOperation) {//Init UI via operation
            initUiViaOperationValues(extras);
        } else {
            primaryUiInit();
        }
    }

    public void primaryUiInit() {
        initDateTimePicker(new Date());
        String dateButtonTitle = getDateButtonTitleByDate(new Date().getTime());
        dateButton.setText(dateButtonTitle);
    }

    public void initUiViaOperationValues(Bundle extras) {
        String dateButtonTitle;
        Operation operation = initOperationFromExtras(extras);
        operation.setOperationDate(new Date(extras.getLong(DATE_KEY)));

        initDateTimePicker(new Date(operation.getOperationDate().getTime()));
        dateButtonTitle = getDateButtonTitleByDate(operation.getOperationDate().getTime());
        dateButton.setText(dateButtonTitle);
        comment.setText(operation.getComment());
        setAmountResultText(operation.getAmount().toString());
        presenter.settingResultText(operation.getAmount());
        selectSpinnerItemMatchesToId(operation.getCategory().getId(), categorySpinnerItemList, categoriesSpinner);
        selectSpinnerItemMatchesToId(operation.getAccountId(), accountSpinnerItemList, accountsSpinner);
    }

    public void selectSpinnerItemMatchesToId(int id, List<SpinnerItem> categorySpinnerItemList, Spinner spinner) {
        for (int i = 0; i < categorySpinnerItemList.size(); i++) {
            if (categorySpinnerItemList.get(i).getId() == id) {
                spinner.setSelection(i);
            }
        }
    }

    public Operation initOperationFromExtras(Bundle extras) {
        Operation operation = (Operation) extras.getSerializable(MainActivity.OPERATION_KEY);
        operation.setAmount(new BigDecimal(getIntent().getExtras().get(MainActivity.AMOUNT_KEY).toString()));
        presenter.setModifyingOperationId(operation.getId());
        return operation;
    }

    public void initSpinnersWithItemLists(List<SpinnerItem> accountSpinnerItemList, List<SpinnerItem> categorySpinnerItemList) {
        ImageSpinnerAdapter accountSpinnerAdapter = new ImageSpinnerAdapter(this, R.layout.image_spinner_item, accountSpinnerItemList,
                R.color.dark_gray, R.color.dark_black);
        accountsSpinner.setAdapter(accountSpinnerAdapter);
        accountsSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            public void onItemSelected(AdapterView<?> parent,
                                       View itemSelected, int selectedItemPosition, long selectedId) {

                accountId = ((SpinnerItem) parent.getItemAtPosition(selectedItemPosition)).getId();
            }

            public void onNothingSelected(AdapterView<?> parent) {
            }
        });

        ImageSpinnerAdapter categoriesSpinnerAdapter = new ImageSpinnerAdapter(this, R.layout.image_spinner_item, categorySpinnerItemList,
                 R.color.dark_gray, R.color.dark_black);
        categoriesSpinner.setAdapter(categoriesSpinnerAdapter);
        categoriesSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            public void onItemSelected(AdapterView<?> parent,
                                       View itemSelected, int selectedItemPosition, long selectedId) {

                categoryId = ((SpinnerItem) parent.getItemAtPosition(selectedItemPosition)).getId();
            }

            public void onNothingSelected(AdapterView<?> parent) {
            }
        });
    }

    @OnClick({R.id.closeOperationButton, R.id.saveRecordButton})
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.closeOperationButton:
                finishActivity();
                break;
            case R.id.saveRecordButton:
                presenter.onButtonSaveClick();
                break;
            default:
        }
    }

    @OnTextChanged({R.id.amountMoneyActivityEditText})
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        if (before > count) presenter.calculatorReset();
    }
}

File: app/src/main/java/com/example/vlad/financemanager/ui/activities/MoneyCalculatorActivity.kt
package com.example.vlad.financemanager.ui.activities

import android.annotation.SuppressLint
import android.app.DatePickerDialog
import android.content.Intent
import android.graphics.Color
import android.graphics.PorterDuff
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.KeyEvent
import android.view.View
import android.widget.AdapterView
import android.widget.Button
import android.widget.DatePicker
import android.widget.ImageButton
import android.widget.Spinner
import android.widget.Toast
import butterknife.OnClick

import com.example.vlad.financemanager.data.database.DatabaseHelper
import com.example.vlad.financemanager.data.mappers.SpinnerItemMapper
import com.example.vlad.financemanager.ui.IMoneyCalculation
import com.example.vlad.financemanager.PresenterMoneyCalculator
import com.example.vlad.financemanager.R
import com.example.vlad.financemanager.data.models.Operation
import com.example.vlad.financemanager.data.models.SpinnerItem
import com.example.vlad.financemanager.ui.adapters.ImageSpinnerAdapter
import com.example.vlad.financemanager.ui.fragments.DatePickerFragment
import kotlinx.android.synthetic.main.activity_money_calculator.*

import java.math.BigDecimal
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date

class MoneyCalculatorActivity : AppCompatActivity(), IMoneyCalculation,
        DatePickerDialog.OnDateSetListener {

    companion object {
        const val DATE_KEY = "date_key"
        private const val DATE_PICKER_TAG = "date picker"
    }

    override val comment: String
        get() {
            return commentMoneyActivityEditText.text.toString()
        }
    override val amount: String
        get() {
            return amountMoneyActivityEditText.text.toString()
        }

    private val mediumDateFormat = SimpleDateFormat("E, dd MMMM")
    private val longDateFormat = SimpleDateFormat("E, MMMM dd, yyyy")

    private lateinit var presenter: PresenterMoneyCalculator
    override var operationDate = Date()
    private lateinit var databaseHelper: DatabaseHelper

    private lateinit var categorySpinnerItemList: List<SpinnerItem>
    private lateinit var accountSpinnerItemList: List<SpinnerItem>

    override var isOperationInput: Boolean = false
    override var accountId: Int = 0
    override var categoryId: Int = 0

    @SuppressLint("ResourceAsColor")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_money_calculator)

        databaseHelper = DatabaseHelper.getInstance(applicationContext)
        presenter = PresenterMoneyCalculator(this)

        calculatorBackButton.setOnLongClickListener {
            presenter.clearNumber()
            true
        }

        initViewsWithClickListeners()
        initAmountEditText()
        initUiViaExtras(intent.extras)
    }

    private fun initViewsWithClickListeners() {
        amountMoneyActivityEditText.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(p0: Editable?) {}

            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (before > count) presenter.calculatorReset()
            }
        })

        closeOperationButton.setOnClickListener { finishActivity() }
        saveRecordButton.setOnClickListener { presenter.onButtonSaveClick() }
        calculatorBackButton.setOnClickListener {
            val pressedButton = it as ImageButton
            presenter.calculatorBtnOnClick(pressedButton.id)
        }
    }

    private fun initAmountEditText() {
        amountMoneyActivityEditText.apply {
            requestFocus()
            setOnKeyListener { _, keyCode, keyEvent ->
                if (!(keyEvent.action == KeyEvent.ACTION_DOWN && keyCode == KeyEvent.KEYCODE_ENTER)) {
                    presenter.calculatorReset()
                }
                false
            }
        }
    }


    private fun initSpinnersItemLists(userId: Int) {
        val accountList = databaseHelper.getAllAccounts(userId)
        accountSpinnerItemList = SpinnerItemMapper.mapAccountsToSpinnerItems(accountList)
        val categoryList = databaseHelper.getAllCategories(userId, isOperationInput)
        categorySpinnerItemList = SpinnerItemMapper.mapCategoryToSpinnerItems(categoryList)
    }

    private fun initToolbar(isIncome: Boolean) {
        setSupportActionBar(calculatorActivityToolbar)
        supportActionBar?.let {
            it.setDisplayShowHomeEnabled(true)
            it.setDisplayHomeAsUpEnabled(true)
        }
        calculatorActivityToolbar.navigationIcon?.setColorFilter(Color.WHITE, PorterDuff.Mode.SRC_ATOP)
        calculatorActivityToolbar.setNavigationOnClickListener { finish() }

        val title = if (isIncome) getString(R.string.income) else getString(R.string.outcome)
        toolbarTitleTextView.text = title
    }

    private fun initDateTimePicker(operationDate: Date) {
        operationDateButton.apply {
            text = mediumDateFormat.format(Date())
        }
        operationDateButton.setOnClickListener {
            DatePickerFragment().apply {
                setCalendar(operationDate)
                show(supportFragmentManager, DATE_PICKER_TAG)
            }
        }
    }

    private fun getDateButtonTitleByDate(dateInMillis: Long): String {
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = dateInMillis
        operationDate = calendar.time

        val dateButtonTitle: String
        dateButtonTitle = if (calendar.get(Calendar.YEAR) != Calendar.getInstance().get(Calendar.YEAR)) {
            longDateFormat.format(calendar.time)
        } else {
            mediumDateFormat.format(calendar.time)
        }

        return dateButtonTitle
    }

    override fun setAmountResultText(resultText: String) {
        amountMoneyActivityEditText.setText(resultText)
        calculationResultTextView.text = resultText
    }

    override fun setCalculatorToZero() {
        calculationResultTextView.text = "0"
    }

    fun calculatorBtnOnClick(view: View) {
        val pressedButton = view as Button
        presenter.calculatorBtnOnClick(pressedButton.id, pressedButton.text.toString())
    }

    override fun finishActivity() {
        finish()
    }

    override fun calculationErrorSignal() {
        Toast.makeText(this, getString(R.string.incorrect_input), Toast.LENGTH_SHORT).show()
    }

    override fun calculationErrorSignal(msg: String) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }

    override fun sendNewOperation(operation: Operation) {
        val amount = operation.amount.toString()
        val extras = Bundle().apply {
            putString(MainActivity.AMOUNT_KEY, amount)
            putParcelable(MainActivity.OPERATION_KEY, operation)
        }
        val intent = Intent().apply {
            putExtras(extras)
        }

        setResult(0, intent)
    }

    override fun onDateSet(view: DatePicker, year: Int, month: Int, dayOfMonth: Int) {
        val calendar = Calendar.getInstance().apply {
            set(Calendar.YEAR, year)
            set(Calendar.MONTH, month)
            set(Calendar.DAY_OF_MONTH, dayOfMonth)
        }

        operationDate = calendar.time
        operationDateButton.text = getDateButtonTitleByDate(operationDate.time)
    }

    private fun initUiViaExtras(extras: Bundle) {
        val userId = extras.getInt(MainActivity.USER_ID_KEY)
        isOperationInput = extras.getBoolean(MainActivity.IS_OPERATION_INCOME)
        val isModifyingOperation = extras.getBoolean(MainActivity.IS_MODIFYING_OPERATION)

        initSpinnersItemLists(userId)
        initSpinnersWithItemLists(accountSpinnerItemList, categorySpinnerItemList)
        initToolbar(isOperationInput)

        if (isModifyingOperation) {//Init UI via operation
            val operation = initOperationFromExtras(extras)
            operation.operationDate = Date(extras.getLong(DATE_KEY))
            presenter.initUiViaOperationValues(operation)
        } else {
            primaryUiInit()
        }
    }

    private fun primaryUiInit() {
        initDateTimePicker(Date())
        val dateButtonTitle = getDateButtonTitleByDate(Date().time)
        operationDateButton.text = dateButtonTitle
    }

    override fun initUiViaOperationValues(operation: Operation) {
        val dateButtonTitle: String = getDateButtonTitleByDate(operation.operationDate.time)

        initDateTimePicker(Date(operation.operationDate.time))
        operationDateButton.text = dateButtonTitle
        commentMoneyActivityEditText.setText(operation.comment)
        setAmountResultText(operation.amount.toString())
        selectSpinnerItemMatchesToId(operation.category.id, categorySpinnerItemList, categorySpinner)
        selectSpinnerItemMatchesToId(operation.accountId, accountSpinnerItemList, accountSpinner)
    }

    private fun selectSpinnerItemMatchesToId(id: Int, categorySpinnerItemList: List<SpinnerItem>, spinner: Spinner?) {
        val spinnerSelectionIndex = categorySpinnerItemList.indexOfFirst { it.id == id }
        spinner?.setSelection(spinnerSelectionIndex)
    }

    private fun initOperationFromExtras(extras: Bundle): Operation {
        val operation = extras.getParcelable(MainActivity.OPERATION_KEY) as Operation
        operation.amount = BigDecimal(extras.get(MainActivity.AMOUNT_KEY)!!.toString())
        presenter.setModifyingOperationId(operation.id)
        return operation
    }

    private fun initSpinnersWithItemLists(accountSpinnerItemList: List<SpinnerItem>, categorySpinnerItemList: List<SpinnerItem>) {
        val accountSpinnerAdapter = ImageSpinnerAdapter(this, R.layout.image_spinner_item, accountSpinnerItemList,
                R.color.dark_gray, R.color.dark_black)
        accountSpinner.apply {
            adapter = accountSpinnerAdapter
            onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>,
                                            itemSelected: View, selectedItemPosition: Int, selectedId: Long) {
                    accountId = (parent.getItemAtPosition(selectedItemPosition) as SpinnerItem).id
                }

                override fun onNothingSelected(parent: AdapterView<*>) {}
            }
        }

        val categoriesSpinnerAdapter = ImageSpinnerAdapter(this, R.layout.image_spinner_item, categorySpinnerItemList,
                R.color.dark_gray, R.color.dark_black)
        categorySpinner.apply {
            adapter = categoriesSpinnerAdapter
            onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
                override fun onItemSelected(parent: AdapterView<*>,
                                            itemSelected: View, selectedItemPosition: Int, selectedId: Long) {
                    categoryId = (parent.getItemAtPosition(selectedItemPosition) as SpinnerItem).id
                }

                override fun onNothingSelected(parent: AdapterView<*>) {}
            }
        }
    }
}
